naked functions

folkertdev opened the pull request:
tracking issue: https://github.com/rust-lang/rust/issues/90957
earlier version: https://github.com/rust-lang/reference/pull/1153

Pending on https://github.com/rust-lang/rust/pull/128004. Once merged, we'll move for stabliziation.

I'm working off of the original PR to the reference, written 2+ years ago when the reference was less strict. I suspect some refinement and editing will be needed here to satisfy the quality requirements of the spec, and update the text to accurately reflect the current state and the exact guarantees/requirements.

Stabilization PR:

- https://github.com/rust-lang/rust/pull/134213

cc @bstrie @traviscross @Amanieu @bjorn3

bstrie committed:
Add documentation for the `naked` function attribute

bstrie committed:
Add `naked` to the attribute index

folkertdev committed:
document `naked_asm!`

folkertdev committed:
mention incompatible attributes

folkertdev committed:
remove unused links

Amanieu commented on src/attributes/codegen.md:
```diff
@@ -51,6 +51,53 @@ r[attributes.codegen.cold]
+No function prologue or epilogue are generated for the attributed function: the contents
+of the `naked_asm!` invocation make up the full body of a naked function.
+
+r[attributes.codegen.naked.call-stack]
+The caller must set up the call stack according to the specified calling convention before
+executing a naked function, even in contexts where setting up the call stack would ordinarily
+be unnecessary, such as when the function is inlined.
+
+An implementation can fulfill this requirement by guaranteeing that naked functions
+are never inlined. However, implementations are not currently required to guarantee that
+naked functions are never inlined.
+
+In the future it may become a requirement for implementations to guarantee that
+naked functions are never inlined; users must not rely on any observable behavior
+that may result from inlining. according to the specified calling convention before
+executing a naked function,
```
I don't think we should be talking about inlining in the reference since that's a specific compiler optimization. Instead we should say what we guarantee:
- The asm code will have a valid call stack and register state on entry as per the signature and calling convention of the function.
- The asm code may not be duplicated by the compiler. This is important for naked functions that define symbols in the assembly code.

Amanieu commented on src/inline-assembly.md:
```diff
@@ -578,9 +587,13 @@ r[asm.options.checks.pure]
 - It is a compile-time error to specify `noreturn` on an asm block with outputs.
 
+r[asm.options.naked_asm-restriction]
+`global_asm!` only supports the `att_syntax` and `raw` options.
```
```suggestion
`naked_asm!` only supports the `att_syntax` and `raw` options.
```

Amanieu commented on src/inline-assembly.md:
```diff
@@ -693,6 +706,44 @@ r[asm.rules.x86-prefix-restriction]
+    Notably it is not the same as an LLVM `undef` which can have a different value every time you read it (since such a concept does not exist in assembly code).
+
+r[asm.naked-rules.reg-not-output]
+- Any callee-saved registers must have the same value upon exiting the asm block as they had on entry, otherwise behavior is undefined.
```
"exiting the asm block" is a bit vague here since you can exit by calling another function. This rule is specifically about the expected state when jumping to the return address provided by the calling convention.

Amanieu commented on src/inline-assembly.md:
```diff
@@ -693,6 +706,44 @@ r[asm.rules.x86-prefix-restriction]
+  - This also applies if the assembly code calls a function which then unwinds.
+
+r[asm.naked-rules.noreturn]
+- Behavior is undefined if execution falls through to the end of the asm block.
```
Maybe add a sub-point that the asm code is expected to provide its own return instruction.

Amanieu commented on src/inline-assembly.md:
```diff
@@ -693,6 +706,44 @@ r[asm.rules.x86-prefix-restriction]
+- The compiler cannot assume that the instructions in the asm are the ones that will actually end up executed.
+  - This effectively means that the compiler must treat the `naked_asm!` as a black box and only take the interface specification into account, not the instructions themselves.
+  - Runtime code patching is allowed, via target-specific mechanisms.
+  - However there is no guarantee that each `naked_asm!` directly corresponds to a single instance of instructions in the object file: the compiler is free to duplicate or deduplicate `naked_asm!` blocks.
```
This is not true, naked functions cannot be duplicated/merged, unlike `asm!`.

chorman0773 replied:
I think that poses problems on SPIR-V or NVPTX targets. I seem to recall that the loader just inlines the functions at the Shader Assembly level.

folkertdev replied:
@chorman0773 just to be sure, you're referring to Amanieu's comment right?

What we do in rustc (or well, we will once that PR makes it through the queue) is that a naked function

```rust
#[naked]
extern "C" fn foo() {}
```

is emitted as something similar to

```rust
core::arch::global_asm!( 
    "foo:",
    "ret"
);

extern "C" {
    fn foo();
}
```

and we're relying on the compiler (rustc, llvm) not duplicating that bit of global assembly. I suspect we already rely on that in general, because global assembly can define symbols today.

If you still think this is problematic, how could we verify the behavior of this new codegen strategy for the targets you list?

Amanieu commented on src/inline-assembly.md:
```diff
@@ -693,6 +706,44 @@ r[asm.rules.x86-prefix-restriction]
+  - However there is no guarantee that each `naked_asm!` directly corresponds to a single instance of instructions in the object file: the compiler is free to duplicate or deduplicate `naked_asm!` blocks.
+
+r[asm.naked-rules.not-exactly-once]
+- You cannot assume that an `naked_asm!` block will appear exactly once in the output binary.
+  The compiler is allowed to instantiate multiple copies of the `naked_asm!` block, for example when the function containing it is inlined in multiple places.
```
Same here.

folkertdev committed:
changes after code review

also maybe make links work?

folkertdev committed:
fix link

folkertdev committed:
add rule on black box/code patching

bjorn3 commented on src/inline-assembly.md:
```diff
@@ -693,6 +706,40 @@ r[asm.rules.x86-prefix-restriction]
+  - Refer to the unsafe code guidelines for the exact rules.
+  - These rules do not apply to memory which is private to the asm code, such as stack space allocated within the asm block.
+
+r[asm.rules.black-box]
```
```suggestion
r[asm.naked-rules.black-box]
```

otherwise there are two rules with the same name. Also I'm surprised this rule isn't already specified globally for all kinds of asm rather than just for function local inline asm.

folkertdev replied:
the impression I get is that `global_asm!` was never properly added: there is no example and the rules are specified for `asm!` and you have to read between the lines to figure out what is relevant for `global_asm!`.

folkertdev committed:
fix rule name

Co-authored-by: bjorn3 <17426603+bjorn3@users.noreply.github.com>

folkertdev committed:
add note about monomorphization

pthariensflame commented on src/inline-assembly.md:
```diff
@@ -61,6 +62,7 @@ option := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn" / "nost
 options := "options(" option *("," option) [","] ")"
 operand := reg_operand / clobber_abi / options
 asm := "asm!(" format_string *("," format_string) *("," operand) [","] ")"
+naked_asm := "asm!(" format_string *("," format_string) *("," operand) [","] ")"
```
```suggestion
naked_asm := "naked_asm!(" format_string *("," format_string) *("," operand) [","] ")"
```

folkertdev committed:
Update src/inline-assembly.md

Co-authored-by: Laine Taffin Altman <alexanderaltman@me.com>

folkertdev committed:
remove the rule on unwinding.

this only applies to `asm!` (even though the rule has been removed there since copying it for naked asm

folkertdev committed:
use `naked_asm!` block to clarify we're talking just about naked assembly

folkertdev committed:
reintroduce a section on unwinding out of `naked_asm!`

folkertdev commented on src/inline-assembly.md:
````diff
@@ -1331,6 +1344,70 @@ r[asm.rules.x86-prefix-restriction]
+
+<!-- #[naked] is currently unstable, tests would fail if this were marked `rust` -->
+```txt
+# #[cfg(target_arch = "x86_64")] {
+#[naked]
+extern "C-unwind" fn naked_function() {
+    unsafe {
+        core::arch::naked_asm!(
+            ".cfi_startproc",
+            "push rbp",
+            ".cfi_def_cfa_offset 16",
+            ".cfi_offset rbp, -16",
+            "mov rbp, rsp",
+            ".cfi_def_cfa_register rbp",
+            "",
+            "call {function}",
+            "",
+            "pop rbp",
+            ".cfi_def_cfa rsp, 8",
+            "ret",
+            ".cfi_endproc",
+            function = sym function_that_panics,
+        )
+    }
+}
+
+extern "C-unwind" fn function_that_panics() {
+    panic!("unwind!");
+}
+# }
````
I don't think I can currently run this because `#[naked]` is still unstable. So reminder to self: update this before merging.

cc @traviscross

folkertdev replied:
Locally I still even get errors like this

```
Couldn't compile the test.
---- inline-assembly.md - Inline_assembly::Options (line 1360) stdout ----
error[E0658]: label operands for inline assembly are unstable
 --> inline-assembly.md:1364:32
  |
6 |       core::arch::asm!("jmp {}", label {
  |  ________________________________^
7 | |         println!();
8 | |     }, options(noreturn));
  | |_____^
  |
  = note: see issue #119364 <https://github.com/rust-lang/rust/issues/119364> for more information
```

with these versions in the `reference` folder (so I don't think there is a rustup override active)

```
> cargo --version
cargo 1.88.0-nightly (d811228b1 2025-04-15)
> rustc --version
rustc 1.88.0-nightly (191df20fc 2025-04-18)
```

That feature has been stable on nightly for a while, so I must be missing something. Anyway, I believe the current code to be correct, and CI will catch it before merge if there are minor issues.

joshtriplett commented on src/inline-assembly.md:
```diff
@@ -2,10 +2,11 @@ r[asm]
 
 r[asm.intro]
-Support for inline assembly is provided via the [`asm!`] and [`global_asm!`] macros.
+Support for inline assembly is provided via the [`asm!`], [`naked_asm!`] and [`global_asm!`] macros.
```
```suggestion
Support for inline assembly is provided via the [`asm!`], [`naked_asm!`], and [`global_asm!`] macros.
```

joshtriplett commented on src/inline-assembly.md:
````diff
@@ -384,8 +390,11 @@ assert_eq!(y, 1);
 ```
 
+r[asm.operand-type.naked_asm-restriction]
+Because `naked_asm!` defines a whole function body, it can only use `sym` and `const` operands.
````
```suggestion
Because `naked_asm!` defines a whole function body and the compiler cannot provide any additional code to handle operands, it can only use `sym` and `const` operands.
```

folkertdev committed:
formatting + update to unsafe(naked)

traviscross committed:
Remove normative cites to unsafe code guidelines doc

The unsafe code guidelines document describes itself currently as
largely abandoned and indicates that those points that have found
consensus are documented in the Reference.  Let's not then normatively
cite that other document.

Let's also remove a line about "these rules do not apply..." where
"these" changed meaning when this line was copied from elsewhere.  In
the original, "these" meant rules about `readonly` and `nomem`,
whereas here, it would have seemed to refer to the rules in the UCG,
and that wouldn't make sense.

traviscross committed:
Sort some lists of things better

traviscross committed:
Make some editorial improvements

traviscross committed:
Annotate naked unwind example and add resources

We demonstrate how to correctly unwind from a naked function.  The
assembler directives and boilerplate for this may be unfamiliar to
people, so let's annotate this example heavily and add links to
further resources.

We'll use the `sysv64-unwind` ABI rather than `C-unwind` just to make
things a bit more unambiguous.

folkertdev commented:
Those final changes look good to me. The feature has been stabilized and CI passes here, so I think that should be everything.

folkertdev commented on src/inline-assembly.md:
```diff
@@ -1362,6 +1384,67 @@ r[asm.rules.preserves_flags]
+
+r[asm.naked-rules.reg-not-output]
+- Any callee-saved registers must have the same value upon return as they had on entry, otherwise behavior is undefined.
+  - Caller-saved registes may be used freely, even if they are not used for the return value.
```
well if the register is used for the return value, any manipulations that you'd perform on that register before then would be overwritten, and hence not observable from the outside.

Changes to other caller-saved registers can be observed by the caller, but the calling convention says that the caller cannot rely on the value in the registers.

So I think this just re-states a property of caller-saved registers.

traviscross committed:
Clarify rules on callee/caller saved registers

For naked functions, let's break out the rules for callee and caller
saved registers and name the rule identifiers more clearly.

For the caller saved registers, let's drop the caveat about "even if
they are not used for the return value", as it doesn't really add
anything to the rule.
